nerfstudio/exporter/texture_utils.py

```diff
@@ -0,0 +1,275 @@
+    uv_indices = torch.stack(
+        torch.meshgrid(
+            torch.arange(num_pixels_w, device=device), torch.arange(num_pixels_h, device=device), indexing="xy"
+        ),
+        dim=-1,
+    )
+    linspace_h = torch.linspace(px_h / 2, 1 - px_h / 2, num_pixels_h, device=device)
+    linspace_w = torch.linspace(px_w / 2, 1 - px_w / 2, num_pixels_w, device=device)
+    uv_coords = torch.stack(
+        torch.meshgrid(linspace_w, linspace_h, indexing="xy"), dim=-1
+    )  # (num_pixels_h, num_pixels_w, 2)
+
+    u_index = torch.div(uv_indices[..., 0], px_per_square_w, rounding_mode="floor")
+    v_index = torch.div(uv_indices[..., 1], px_per_square_h, rounding_mode="floor")
+    square_index = v_index * squares_per_side_w + u_index
+    u_offset = uv_indices[..., 0] % px_per_square_w
+    v_offset = uv_indices[..., 1] % px_per_square_h
+    lower_right = (u_offset + v_offset) >= (px_per_square_w - 2)
+    triangle_index = square_index * 2 + lower_right
+    triangle_index = torch.clamp(triangle_index, min=0, max=len(faces) - 1)
+
+    nearby_uv_coords = texture_coordinates[triangle_index]  # (num_pixels_h, num_pixels_w, 3, 2)
+    nearby_vertices = vertices[faces[triangle_index]]  # (num_pixels_h, num_pixels_w, 3, 3)
+    nearby_normals = vertex_normals[faces[triangle_index]]  # (num_pixels_h, num_pixels_w, 3, 3)
+
+    # compute barycentric coordinates
+    v0 = nearby_uv_coords[..., 0, :]  # (num_pixels, num_pixels, 2)
+    v1 = nearby_uv_coords[..., 1, :]  # (num_pixels, num_pixels, 2)
+    v2 = nearby_uv_coords[..., 2, :]  # (num_pixels, num_pixels, 2)
+    p = uv_coords  # (num_pixels, num_pixels, 2)
+    epsilon = 1e-8
+    area = get_parallelogram_area(v2, v0, v1) + epsilon  # 2 x face area.
+    w0 = get_parallelogram_area(p, v1, v2) / area
+    w1 = get_parallelogram_area(p, v2, v0) / area
+    w2 = get_parallelogram_area(p, v0, v1) / area
+
+    # TODO: clean this up
+    # compute the length of the rays we want to render
+    # we can make a good guess for this by looking at the average distance between vertices on the triangle faces
+    neary_vertices_dev = nearby_vertices.to(device)
+    offset = torch.mean(torch.norm(neary_vertices_dev[..., 0, :] - neary_vertices_dev[..., 1, :], dim=-1))
+
+    origins = (
+        nearby_vertices[..., 0, :] * w0[..., None]
+        + nearby_vertices[..., 1, :] * w1[..., None]
+        + nearby_vertices[..., 2, :] * w2[..., None]
+    ).to(device)
+    directions = -(
+        nearby_normals[..., 0, :] * w0[..., None]
+        + nearby_normals[..., 1, :] * w1[..., None]
+        + nearby_normals[..., 2, :] * w2[..., None]
+    ).to(device)
+    # normalize the direction vector to make it a unit vector
+    directions = torch.nn.functional.normalize(directions, dim=-1)
+
+    origins = origins - offset * directions
+    pixel_area = torch.ones_like(origins[..., 0:1])
+    camera_indices = torch.zeros_like(origins[..., 0:1])
+    nears = torch.zeros_like(origins[..., 0:1])
```

Comment:
I forget, will the near plane get clipped to a larger value elsewhere in the code?

Reply:
Good catch, this was a bug. I updated the SceneCollider classes.